options.go

```diff
@@ -41,6 +48,48 @@ func CaseInsensitive(tokens ...string) Option {
 	}
 }
 
+func UseCustom(parseFn interface{}) Option {
+	errorType := reflect.TypeOf((*error)(nil)).Elem()
+	peekingLexerType := reflect.TypeOf((*lexer.PeekingLexer)(nil))
+	return func(p *Parser) error {
+		parseFnVal := reflect.ValueOf(parseFn)
+		parseFnType := parseFnVal.Type()
+		if parseFnType.Kind() != reflect.Func {
+			return fmt.Errorf("production parser must be a function (got %s)", parseFnType)
+		}
+		if parseFnType.NumIn() != 1 || parseFnType.In(0) != reflect.TypeOf((*lexer.PeekingLexer)(nil)) {
+			return fmt.Errorf("production parser must take a single parameter of type %s", peekingLexerType)
+		}
+		if parseFnType.NumOut() != 2 {
+			return fmt.Errorf("production parser must return exactly two values: the parsed production, and an error")
+		}
+		if parseFnType.Out(0).Kind() != reflect.Interface {
+			return fmt.Errorf("production parser's first return must be an interface type")
+		}
+		if parseFnType.Out(1) != errorType {
+			return fmt.Errorf("production parser's second return must be %s", errorType)
+		}
+		prodType := parseFnType.Out(0)
+		p.customDefs = append(p.customDefs, customDef{prodType, parseFnVal})
+		return nil
+	}
+}
+
+func UseUnion[T any](members ...T) Option {
```

Comment:
So beautiful, brings a tear to my eye 🥲

Reply:
Also this will need documentation.

Reply:
Documented, also renamed to `ParseUnion` to be consistent with the `ParseTypeWith` option.